setup.py

```diff
@@ -1,4 +1,12 @@
 from setuptools import setup
 
+# version.py is created when building with SCons.
+# Otherwise, we create it ourselves.
+version = "0.1.0"
+with open("./python/lsst/alert/packet/version.py", "w") as f:
+    print(f"""
+__all__ = ("__version__", )
```

Comment:
No need for `__all__` if `__version__` is the only symbol in the file, right?

Reply:
Again, this is purely to match what sconsUtils was doing.

Reply:
But SCons produces different results anyway, right? I get this:

```python
# -------- This file is automatically generated by LSST's sconsUtils -------- #
__version__ = 'w.2020.25-7-g6c29e90'
__repo_version__ = 'w.2020.25-7-g6c29e90'
__fingerprint__ = '6c29e90ec6ea5fb61328e5cf8e6859a9ccad4e4c'
__dependency_versions__ = {
}
__all__ = (
    '__version__',
    '__repo_version__',
    '__fingerprint__',
    '__dependency_versions__',
)
```

What I did:
```bash
$ git checkout tickets/DM-25642
Branch 'tickets/DM-25642' set up to track remote branch 'tickets/DM-25642' from 'origin'.
Switched to a new branch 'tickets/DM-25642'

$ git clone git@github.com:lsst/sconsUtils.git ../sconsUtils
Cloning into '../sconsUtils'...
remote: Enumerating objects: 169, done.
remote: Counting objects: 100% (169/169), done.
remote: Compressing objects: 100% (101/101), done.
remote: Total 4480 (delta 98), reused 107 (delta 52), pack-reused 4311
Receiving objects: 100% (4480/4480), 832.79 KiB | 3.64 MiB/s, done.
Resolving deltas: 100% (2150/2150), done.

$ ln -s $PWD/../sconsUtils ./python/lsst
$ scons                                                                 
scons: Reading SConscript files ...
EUPS integration: disabled
Unable to import eups; guessing flavor
Setting up environment to build package 'alert_packet'.
scons: done reading SConscript files.
scons: Building targets ...
scons: Nothing to be done for `python'.
scons: Nothing to be done for `examples'.
scons: Nothing to be done for `doc'.
scons: done building targets.

$ cat python/lsst/alert/packet/version.py
```

Reply:
Ok. I misremembered that `__all__`. I'm just in the habit of always adding one.  The important thing was the creation of the version.py file since that needs to exist to support scons and setup.py.

Reply:
Why does that file need to exist to support setup.py? I'm able to run setup.py without that file in this branch.

Do you mean that it's needed to support the change to `lsst/alert/__init__.py` which adds `from .version import *`? That's true, but I don't know why we need that line - is that to support something else happening in SCons?

Reply:
If you aren't using sconsUtils then the only requirement here is that `lsst.alert.packet.__version__` reports the version string (we need that convention for provenance tracking in pipelines).

Reply:
The issue at the moment is that we're not using setuptools in stack builds. Presumably the right way around that is to fix DM-25281, rather than to drop to sconsUtils.

That still makes me wonder:

- Can we use setuptools_scm in the stack build by just listing it in `setup_requires`, or will that upset something?
- Can we make the version number generated by setuptools_scm consistent with that generated by `pkgautoversion` and used `eups distrib`? It looks like the default setuptools_scm scheme is close-but-not-quite-there.

Did i mention rabbit holes?

Reply:
@jdswinbank, I don't understand. I might not be clear on what you mean by "not using setuptools in stack builds;" this package is successfully getting used by `ap_association`, even though it doesn't have a `__version__` variable and it doesn't have a `SConstruct. Is it broken today in some way because of that, and that's what you're trying to fix? I am probably missing lotso f context - I don't know what `pkgautoversion` is or how `eups distrib` works in any way, so I don't know what the connection is.

I agree about rabbit holes, I'm lost on why you need to change versioning to add sphinx documentation.

Reply:
Isn't it a python convention to support `__version__`?  It's definitely used by the pipeline provenance system to dump package versions.

Reply:
> Isn't it a python convention to support __version__? It's definitely used by the pipeline provenance system to dump package versions.

Yeah, I think that's https://www.python.org/dev/peps/pep-0396/. Exposing a variable with that name is mentioned in the setuptools_scm docs (https://github.com/pypa/setuptools_scm#retrieving-package-version-at-runtime).

I'm just unclear on how it's related to the sphinx documentation stuff.

Reply:
Sorry for not being clear! There are two issues here that I think we need to resolve.

First, the documentation build system will access `lsst.alert.packet.__version__`, so we need to make sure that it exists (even though I agree it isn't required by ap_association). Otherwise, we'll end up with something like:

```
[lsst@lsst-docker-w_latest alert_packet]$ package-docs clean && package-docs build
Running Sphinx v1.7.9

Traceback (most recent call last):
  File "/opt/lsst/software/stack/conda/miniconda3-py37_4.8.2/envs/lsst-scipipe-1a1d771/lib/python3.7/site-packages/sphinx/config.py", line 161, in __init__
    execfile_(filename, config)
  File "/opt/lsst/software/stack/conda/miniconda3-py37_4.8.2/envs/lsst-scipipe-1a1d771/lib/python3.7/site-packages/sphinx/util/pycompat.py", line 150, in execfile_
    exec_(code, _globals)
  File "conf.py", line 13, in <module>
AttributeError: module 'lsst.alert.packet' has no attribute 'version'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/lsst/software/stack/conda/miniconda3-py37_4.8.2/envs/lsst-scipipe-1a1d771/lib/python3.7/site-packages/documenteer/sphinxrunner.py", line 84, in run_sphinx
    warningiserror, tags, verbosity, jobs)
  File "/opt/lsst/software/stack/conda/miniconda3-py37_4.8.2/envs/lsst-scipipe-1a1d771/lib/python3.7/site-packages/sphinx/application.py", line 163, in __init__
    confoverrides or {}, self.tags)
  File "/opt/lsst/software/stack/conda/miniconda3-py37_4.8.2/envs/lsst-scipipe-1a1d771/lib/python3.7/site-packages/sphinx/config.py", line 167, in __init__
    raise ConfigError(CONFIG_ERROR % traceback.format_exc())
sphinx.errors.ConfigError: There is a programable error in your configuration file:

Traceback (most recent call last):
  File "/opt/lsst/software/stack/conda/miniconda3-py37_4.8.2/envs/lsst-scipipe-1a1d771/lib/python3.7/site-packages/sphinx/config.py", line 161, in __init__
    execfile_(filename, config)
  File "/opt/lsst/software/stack/conda/miniconda3-py37_4.8.2/envs/lsst-scipipe-1a1d771/lib/python3.7/site-packages/sphinx/util/pycompat.py", line 150, in execfile_
    exec_(code, _globals)
  File "conf.py", line 13, in <module>
    version=lsst.alert.packet.version.__version__))
AttributeError: module 'lsst.alert.packet' has no attribute 'version'
```
(We might consider edting `doc/conf.py` to not access `__version__`, but since — as Tim says — it ought to exist, and since that would make us diverge from the [standard template](https://github.com/lsst/templates/blob/master/project_templates/stack_package/example/doc/conf.py),  I don't like that idea much).

Secondly, we want to get on top of the proliferation of version numbers. At present, we have two: `0.1.0`, in `setup.cfg`, and `master-gaf23745059` (which is generated by [`pkgautoversion`](https://github.com/RobertLuptonTheGood/eups/blob/master/bin/pkgautoversion) based on the output of `git describe`, and is being used by EUPS). If we add `lsst.alert.packet.__version__`, we might end up with three, which would not be progress.

I hope that's a clear(er) statement of the problem. Now, how do we resolve it? My thoughts, which may well be wrong:

- I agree with @spenczar that we want a single source of truth for version information.
- It _may_ be possible to disentangle `pkgautoversion` from EUPS and set the version some other way, but I don't relish the prospect. That implies to me that we want to use something based on git (and, ideally, the output of `git describe`) as that universal source of versioning truth.
- setuptools_scm looks like it will set a version string based on `git describe`, but it's not clear it exactly matches the format produced by `pkgautoversion`. It's presumably possible to tweak it, but I've not yet tried to understand how.
- If setuptools_scm can be made to generate an appropriate version string, then we need to make sure we actually use it when building packages for EUPS (rather than the hack that I put in a few weeks ago, and which we're supposed to remove on DM-25281).

Does that make more sense?

Reply:
Also on pypi you need a version that is semantic don't you? Can setuptools_scm generate developer version numbers that use the git tag so that if you tag it 1.0.0 it reports 1.0.0 but if you have a post 1.0.0 commit it uses the dev form? And does setuptools_scm embed the version into the package that you upload to PyPI since you won't be using git then?

Reply:
setuptools_scm says that by default it'll do:

> no distance and clean:
> `{tag}`
> distance and clean:
> `{next_version}.dev{distance}+{scm letter}{revision hash}`

I think this is _almost_ right — if we have a numeric tag (`0.1.0`), it'll just use it, and if we don't, it'll generate something based on the known tags and the git hash. Unfortunately it inserts `.dev` instead of `-`, and it uses  `next_version` which “is calculated by adding 1 to the last numeric component of the tag”, neither of which match `pkgautoversion`. It might be possible to persuade it to do do the “right” thing with some effort, though.

Reply:
Yes, @jdswinbank, that helps a ton, thank you.

If we agree that we want git to be the source of truth via tags, and we don't want to change `eups` or `pkgautoversion`, then the only remaining thing is to implement `pkgautoversion` in Python. It's unfortunate that `pkgautoversion` doesn't use [PEP 440](https://www.python.org/dev/peps/pep-0440/) in its generation of version srings - that's the format setuptools_scm is using - but we may just have to make do.

I am not sure that the way to this is to use `setuptool_scm` at all. We might be better off implementing `pkgautoversion` from scratch since it's kind of idiosyncratic.

Alternatively, we can depend directly upon `pkgautoversion` in build processes, and shell out to it from `setup.py`. It'll be pretty inscrutable to almost anyone, but at least it's consistent.

Reply:
you can't use eups in the pypi package.

Reply:
Okay, I guess I give up on this. Go with whatever you guys think is best, I'm ducking out.